diff --git a/src/vulkan/vk-acceleration-structure.cpp b/src/vulkan/vk-acceleration-structure.cpp
index a6b037e..3a6bce0 100644
--- a/src/vulkan/vk-acceleration-structure.cpp
+++ b/src/vulkan/vk-acceleration-structure.cpp
@@ -285,7 +285,7 @@ Result AccelerationStructureBuildDescConverter::convert(
                 geometry.geometry.triangles.pNext = &ommData;
             }
 
-            primitiveCounts[i] = max(triangles.vertexCount, triangles.indexCount) / 3;
+            primitiveCounts[i] = (triangles.indexBuffer ? triangles.indexCount : triangles.vertexCount) / 3;
         }
 
         buildInfo.type = VK_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL_KHR;
